### PR TITLE
Rename build/ to m4/

### PR DESCRIPTION
**Pull request by JosiahWI**
This fixes an issue on case-insensitive filesystems where the BUILD item in .gitignore causes the build/ directory to be ignored in git. It also opens up build/ for use as an out of source build directory, which is a very common idiom for CMake users.

### CODE REVIEW

**Review by bneradt — DISMISSED** [review]

File: .gitignore [hunk 1]
```diff
@@ -37,12 +37,12 @@ Makefile.in
...
-build/ltversion.m4
-build/lt~obsolete.m4
+m4/_aux/
+m4/libtool.m4
+m4/ltoptions.m4
+m4/ltsugar.m4
+m4/ltversion.m4
+m4/lt~obsolete.m4
 
 BUILD
```
[the remark this answers is not in the document]
Comment (JosiahWI) [reply .gitignore:47]: Too slow.

**Review by bneradt — COMMENTED** [review]
Looks good. I'll let a non-Yahoo committer give you an approval.

**Comment by moonchen** [comment]
Please also update README.md to reflect the change in directory structure.

**Comment by JosiahWI** [comment]
Rebased on master.

File: m4/tcmalloc.m4 [hunk 2]
```diff
@@ -0,0 +1,56 @@
+dnl -------------------------------------------------------- -*- autoconf -*-
```
Comment (bryancall) [inline m4/tcmalloc.m4:1]: Why was this file added?   There was a PR to remove tcmalloc from the build.

File: example/plugins/c-api/Makefile.am [hunk 3]
```diff
@@ -14,8 +14,8 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
-include $(top_srcdir)/build/plugins.mk
-include $(top_srcdir)/build/tidy.mk
+include $(top_srcdir)/m4/plugins.mk
+include $(top_srcdir)/m4/tidy.mk
```
Comment (cmcfarlen) [inline example/plugins/c-api/Makefile.am:18]: Should we put the makefile includes in `$(top_srcdir)/mk`?

Comment (JosiahWI) [reply example/plugins/c-api/Makefile.am:18]: Great suggestion.

**Comment by JosiahWI** [comment]
[ci approve autest]

**Comment by JosiahWI** [comment]
[approve ci autest]

File: README.md [hunk 4]
```diff
@@ -45,6 +44,7 @@ trafficserver ............. Top src dir
 ├── lib ...................
 │   ├── records ........... Library for config files
 │   └── yamlcpp ........... Library for YAML of C++
+├── m4 .................... Custom macros for configure.ac
```
Comment (bryancall) [inline README.md:47]: mk directory?

**Comment by bneradt** [comment]
[approve ci fedora]

File: ci/rat-regex.txt [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ^BUILDS$
 ^autom4te\.cache$
-^build$
+^m4$
```
Comment (cmcfarlen) [inline ci/rat-regex.txt:3]: Do we want rat to check our m4 files?  I wonder if this line should just be removed? @bneradt do you have any input?

Comment (JosiahWI) [reply ci/rat-regex.txt:3]: RAT checks our CMake scripts. This seems like an analogous case for automake. Just my two cents.

Comment (bneradt) [reply ci/rat-regex.txt:3]: Probably good to check them. Let's do that as a separate PR though in case something goes wrong or we change our mind so it would be easy to revert.

**Comment by bneradt** [comment]
[approve ci autest]

**Comment by jpeach** [comment]
@JosiahWI Can you also remove "BUILD" from the gitignore? It looks like @zwoop added it in https://github.com/apache/trafficserver/pull/8428, but it looks accidental to me.

**Comment by bneradt** [comment]
[approve ci]

**Comment by cmcfarlen** [comment]
[approve ci ubuntu]

**Comment by bneradt** [comment]
[approve ci cmake]

**Review by cmcfarlen — APPROVED** [review]
I feel like enough time has passed for someone to object to this.  Approved!
